crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2513,18 +2513,32 @@ impl<'db> TypeInferenceBuilder<'db> {
             .ok_or(ModuleNameResolutionError::UnknownCurrentModule)?;
         let mut level = level.get();
         if module.kind().is_package() {
-            level -= 1;
+            level = level.saturating_sub(1);
         }
+
         let mut module_name = module.name().clone();
-        for _ in 0..level {
-            module_name = module_name
-                .parent()
-                .ok_or(ModuleNameResolutionError::TooManyDots)?;
+        let tail = tail
+            .map(|tail| ModuleName::new(tail).ok_or(ModuleNameResolutionError::InvalidSyntax))
+            .transpose()?;
```

Comment:
We might be eagerly creating a `ModuleName` from the `tail` here (which might be expensive) in some unnecessary situations... this could be a lazy closure rather than an eager assignment? But what you have is also probably fine

Reply:
The only case where we unnecessarily create a `ModuleName` here is if the import is invalid. Considering that this should (hopefully) be the exception, allocating should be fine.